teuthology/task/ceph_ansible.py

```diff
@@ -113,18 +188,157 @@ def generate_hosts_file(self):
+                                                content=hosts_stringio.read().strip())
         self.generated_inventory = True
 
+    def begin(self):
+        super(CephAnsible, self).begin()
+        self.execute_playbook()
+
+    def _write_hosts_file(self, prefix, content):
+        """
+        Actually write the hosts file
+        """
+        hosts_file = NamedTemporaryFile(prefix=prefix,
+                                        delete=False)
+        hosts_file.write(content)
+        hosts_file.flush()
+        return hosts_file.name
+
+    def wait_for_ceph_health(self):
+        with contextutil.safe_while(sleep=15, tries=6,
+                                    action='check health') as proceed:
+            (remote,) = self.ctx.cluster.only('mon.a').remotes
+            remote.run(args=['sudo', 'ceph', 'osd', 'tree'])
+            remote.run(args=['sudo', 'ceph', '-s'])
+            log.info("Waiting for Ceph health to reach HEALTH_OK \
+                        or HEALTH WARN")
+            while proceed():
+                out = StringIO()
+                remote.run(args=['sudo', 'ceph', 'health'], stdout=out)
+                out = out.getvalue().split(None, 1)[0]
+                log.info("cluster in state: %s", out)
+                if out in ('HEALTH_OK', 'HEALTH_WARN'):
+                    break
+
     def get_host_vars(self, remote):
         extra_vars = self.config.get('vars', dict())
         host_vars = dict()
         if not extra_vars.get('osd_auto_discovery', False):
-            host_vars['devices'] = get_scratch_devices(remote)
+            roles = self.ctx.cluster.remotes[remote]
+            dev_needed = len([role for role in roles
+                              if role.startswith('osd')])
+            host_vars['devices'] = get_scratch_devices(remote)[0:dev_needed]
         if 'monitor_interface' not in extra_vars:
             host_vars['monitor_interface'] = remote.interface
         if 'public_network' not in extra_vars:
             host_vars['public_network'] = remote.cidr
         return host_vars
 
+    def run_rh_playbook(self):
+        ceph_installer = self.ceph_installer
+        args = self.args
+        ceph_installer.run(args=['cp', '-R',
+                                 '/usr/share/ceph-ansible', '.'])
+        ceph_installer.put_file(self.inventory, 'ceph-ansible/inven.yml')
+        ceph_installer.put_file(self.playbook_file, 'ceph-ansible/site.yml')
+        # copy extra vars to groups/all
+        ceph_installer.put_file(self.gvars, 'ceph-ansible/group_vars/all')
+        # print for debug info
+        ceph_installer.run(args=('cat', 'ceph-ansible/inven.yml'))
+        ceph_installer.run(args=('cat', 'ceph-ansible/site.yml'))
+        ceph_installer.run(args=('cat', 'ceph-ansible/group_vars/all'))
+        out = StringIO()
+        str_args = ' '.join(args)
+        ceph_installer.run(args=['cd', 'ceph-ansible', run.Raw(';'),
```

Comment:
This would be nicer formatted like:

```
ceph_installer.run(
    args=[
        'cd', 
        'ceph-ansible',
        run.Raw(';'),
        run.Raw(str_args)
    ],
    timeout=4200,
    check_status=False,
    stdout=out,
)
```